File: android/src/main/java/com/lightningkite/kwift/location/Geocoding.actual.kt
package com.lightningkite.kwift.location

import android.location.Address
import android.location.Geocoder
import com.lightningkite.kwift.delay
import com.lightningkite.kwift.post
import com.lightningkite.kwift.location.GeoCoordinate
import com.lightningkite.kwift.views.ViewDependency

fun ViewDependency.geocode(
    coordinate: GeoCoordinate,
    onResult: (List<GeoAddress>) -> Unit
) {
    if(coordinate.latitude == 0.0 && coordinate.longitude == 0.0){
        onResult(listOf())
        return
    }
    Thread {
        val result = Geocoder(context)
            .getFromLocation(coordinate.latitude, coordinate.longitude, 1)
        post {
            onResult(result)
        }
    }.start()
}

fun ViewDependency.geocode(
    address: String,
    onResult: (List<GeoAddress>) -> Unit
) {
    if(address.isEmpty()){
        onResult(listOf())
        return
    }
    Thread {
        val result = Geocoder(context)
            .getFromLocationName(address, 1)
        post {
            onResult(result)
        }
    }.start()
}

File: ios/Khrysalis/com/lightningkite/khrysalis/location/Geocoding.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation
import CoreLocation

public extension ViewDependency {

    //--- ViewDependency.geocode(GeoCoordinate, (List<GeoAddress>)->Unit)
    func geocode(_ latLng: GeoCoordinate, _ onResult: @escaping (Array<GeoAddress>)->Void) {
        geocode(latLng: latLng, onResult: onResult)
    }
    func geocode(latLng: GeoCoordinate, onResult: @escaping (Array<GeoAddress>)->Void) {
        CLGeocoder().reverseGeocodeLocation(CLLocation(latitude: latLng.latitude, longitude: latLng.longitude)){ marks, error in
            onResult(marks?.map { translate(mark: $0) } ?? [])
        }
    }

    //--- ViewDependency.geocode(String, (List<GeoAddress>)->Unit)
    func geocode(_ address: String, _ onResult: @escaping (Array<GeoAddress>)->Void) {
        geocode(address: address, onResult: onResult)
    }
    func geocode(address: String, onResult: @escaping (Array<GeoAddress>)->Void) {
        CLGeocoder().geocodeAddressString(address){ marks, error in
            onResult(marks?.map { translate(mark: $0) } ?? [])
        }
    }
}

//--- translate
private func translate(mark: CLPlacemark) -> GeoAddress {
    return GeoAddress(mark.location?.coordinate.toKhrysalis(), mark.name, mark.street, mark.subLocality, mark.locality, mark.subAdministrativeArea, mark.administrativeArea, mark.country, mark.postalCode)
}

private extension CLPlacemark {
    
    var street: String? {
        if let x = subThoroughfare, let y = thoroughfare {
            return "\(x) \(y)"
        }
        return nil
    }
}
